Add HTTP API with managing stations

platinumthinker opened the pull request:
- get list of stations with info:
  **/api/v2/station** : 
```
  [
  {
    "wtp-id": "wtp-lede",
    "radios": [
      {
        "radio-id": 2,
        "stations": [
          {
            "mac": "50-55-27-FE-61-10",
            "rssi": -71,
            "vlan": 0,
            "rx-packets": 6,
            "tx-packets": 0,
            "rx-bytes": 364,
            "tx-bytes": 0
          }
        ]
      },
      {
        "radio-id": 1,
        "stations": [
          {
            "mac": "84-3A-4B-33-4A-3C",
            "rssi": -62,
            "vlan": 0,
            "rx-packets": 1774,
            "tx-packets": 0,
            "rx-bytes": 79410,
            "tx-bytes": 0
          }
        ]
      }
    ]
  }
]
```
 - **Delete station connected to wtp**:
    /api/v2/wtp/WTP_ID/station/STATION_MAC

platinumthinker committed:
Add HTTP API v2 with getting stations and delete stations

Add endpoint /api/v2/station for getting information about stations,
grouped by WTP and contains MAC, RSSI, VLAN and RX/TX statistics

RoadRunnr commented on src/capwap_ac.erl:
```diff
@@ -827,7 +827,8 @@ handle_capwap_data(DataPath, WTPDataChannelAddress, Header, true,
 		   Header = #capwap_header{
-			       flags = Flags, radio_mac = RecvRadioMAC},
+			       flags = Flags, radio_mac = RecvRadioMAC,
+                   wireless_spec_info = WSI},
```
white space and indention is wrong

platinumthinker replied:
fixed

RoadRunnr commented on src/capwap_http_api_handler.erl:
```diff
@@ -81,11 +87,54 @@ handle_request_text(Req, State) ->
+         end},
+         {sta_id, fun(Station) ->
+            BinMAC = capwap_tools:mac_to_hex(Station),
+            io:format("paesed ~p~n", [Station]),
```
debug leftover?

platinumthinker replied:
yep. deleted

RoadRunnr commented:
indention and formatting is inconsistent. For Erlang files the default layout of [emacs.el](http://erlang.org/doc/apps/tools/erlang_mode_chapter.html) is the rule.

Please make sure that you use that style in all new code. Sometimes it might not be a bad idea to have  a separate commit that aligns old/existing code with that style first.

platinumthinker commented:
> indention and formatting is inconsistent. For Erlang files the default layout of [emacs.el](http://erlang.org/doc/apps/tools/erlang_mode_chapter.html) is the rule.
> 
> Please make sure that you use that style in all new code. Sometimes it might not be a bad idea to have a separate commit that aligns old/existing code with that style first.

What the style using in the project? I can't find any description - only Emacs plugin.. but i using another text editor.

platinumthinker commented:
@RoadRunnr so, i tried to fix code style (disable replacing tab to space, set tabwith=8)

RoadRunnr commented on src/capwap_tools.erl:
```diff
@@ -31,6 +32,25 @@ format_mac(<<A:8, B:8, C:8, D:8, E:8, F:8>>) ->
 format_mac(MAC) ->
     flat_format("~w", MAC).
 
+mac_to_hex(MAC) when is_binary(MAC)->
```
for to complicated, here is a much simple version: https://github.com/travelping/capwap/blob/1af16c4c354c797113911a236245ccb7c3f92713/src/capwap.escript#L127

RoadRunnr commented on src/capwap_http_api_handler.erl:
```diff
@@ -20,408 +20,459 @@
 hex2dec(C) when C >= $A andalso C =< $F -> C - $A + 10;
 hex2dec(C) when C >= $0 andalso C =< $9 -> C - $0.
 
-mac_to_bin(MAC) ->
```
why did you remove this and replaced it with a much more complicated version (mac_to_hex)?

RoadRunnr commented on src/capwap_http_api_handler.erl:
```diff
@@ -20,408 +20,459 @@
+delete_resource(Req, State) ->
+	handle_request_json(Req, State).
+
+resource_exists(Req = #{bindings := Vars}, State) ->
```
this function makes my head spin, drop the fold and use a recursive function instead, please.

RoadRunnr replied:
this was not resolved, the logic in this function is still expressed in a way that is way to complicated.

HINT: If you have more that 2 level of indented block, your should start thinking about rewriting. Beyond 4 it almost always should be rewritten (here we have foldl->fun->case->case ... that is definitely to much to make sense).

RoadRunnr commented on src/capwap_http_api_v2_handler.erl:
```diff
@@ -0,0 +1,103 @@
+%%%===================================================================
+
+handle(<<"GET">>, [<<"station">>], Req, State) ->
+	Stations0 = capwap_station_reg:list_stations(),
```
this is mostly a copy of capwap:list_stations/0, use that function instead

RoadRunnr commented on src/capwap_http_api_v2_handler.erl:
```diff
@@ -0,0 +1,103 @@
+	ResRadios = maps:fold(fun(RadioId, STAs0, Acc) ->
+				[[{'radio-id', RadioId},
+				  {stations, STAs0}] | Acc]
+			      end, [], Radios), [[{'wtp-id', WTP},
```
line break after the end of the fold function is missing, makes it very hard to read

RoadRunnr commented on src/capwap_http_api_v2_handler.erl:
```diff
@@ -0,0 +1,103 @@
+					[WTP|_] = capwap_wtp_reg:get_commonname(K),
+				{WTP, V}
+			     end, orddict:to_list(Stations1)),
+	Res = lists:foldl(fun format_wtps/2, [], Stations),
```
a lists:maps instead of the foldl should be simpler

RoadRunnr commented on src/capwap_http_api_v2_handler.erl:
```diff
@@ -0,0 +1,103 @@
+		_ -> undefined
+	end.
+
+update_radio(Id, Acc, Station) ->
```
can be done much simpler with:
```maps:update_with(Id, fun(Stations) -> [Station|Stations] end, [Station], Acc).```

If you use the `cut` parse_transform this even becomes:
```maps:update_with(Id, [Station|_] end, [Station], Acc).```

RoadRunnr commented on src/capwap_packet.erl:
```diff
@@ -38,7 +38,14 @@ decode(Type, <<0:4, 0:4,
     <<RestHeader0:RestHeaderLen/bytes, PayLoad/binary>> = Rest,
     {RadioMAC, RestHeader1} = extract_header(M, RestHeader0),
-    {WirelessSpecInfo, _} = extract_header(W, RestHeader1),
+    WirelessSpecInfo = case extract_header(W, RestHeader1) of
```
this should be done in a helper function

RoadRunnr commented on src/capwap_http_api_v2_handler.erl:
```diff
@@ -0,0 +1,92 @@
+		_ -> undefined
+	end.
+
+update_radio(Id, Acc, Station) ->
```
typical argument order for function like this is (Key, NewValue, List)

RoadRunnr commented:
> > indention and formatting is inconsistent. For Erlang files the default layout of [emacs.el](http://erlang.org/doc/apps/tools/erlang_mode_chapter.html) is the rule.
> > Please make sure that you use that style in all new code. Sometimes it might not be a bad idea to have a separate commit that aligns old/existing code with that style first.
> 
> What the style using in the project? I can't find any description - only Emacs plugin.. but i using another text editor.

The style given by the Emacs plugin.

There are other editors besides emacs???? :grimacing:

This http://www.erlang.se/doc/programming_rules.shtml seems to good guide. It doesn't say too much about the indentation, but a whole lot about other aspects of the code structure.

This is also nice https://www.ejabberd.im/Guidelines%20for%20ejabberd%20developers/index.html (except for the TabVsSpace thingy)

Another great guide https://github.com/inaka/erlang_guidelines (again, except for SpaceVsTab thing)

RoadRunnr commented on src/ieee80211_station.erl:
```diff
@@ -149,16 +157,23 @@ init([AC, ClientMAC, StationCfg = #station_config{bss = RadioMAC}]) ->
+    Data1 = update_rssi(FrameInfo, Data),
+    {ok, initial_state(Data1), Data1}.
+
+handle_event(Type, {'802.11', FrameType, DA, SA, BSS, FromDS, ToDS, Frame, FrameInfo}, State,
```
I have to think about a different solution a bit more. But I can already tell you, this change is not acceptable, **never**, **ever**, **ever**. :rage:

platinumthinker replied:
@RoadRunnr i can change every handle_event to catch FrameInfo without wrapping handle_event

Florz replied:
@platinumthinker thanks for all the changes! I gave it a quick try in ergw-capwap-node. `/api/v2/station` worked fine. However wtp-id looked pretty messed up -> `"wtp-id":"���w�7"`. Do I need to configure this separately somewhere?
Using this in a delete request to /`api/v2/wtp/���w�7/station/STATION_MAC` showed no result. Appending or deleting characters from the `WTP_ID` caused no error messages. Setting a random ID caused a message like -> `{"not_found":{"wtp_id":"01234"}}`. I guess this might be related to the strange ID.

The delete command in the version one API seems to work somewhat. I am able do remove  a device from the listing in the API. The data connection also breaks however the device stays associated to the WTP. Right now I tested this with one WTP only. I might repeat this and see if it makes a difference with multiple WTPs.

platinumthinker replied:
@Florz, Hello! What wtp_id you have? with unicode symbols?

Florz replied:
@platinumthinker yes, I saw exactly the one I pasted here.

platinumthinker replied:
@Florz, i mean what name expected? I tested with one wtp with name "wtp-lede" and all works fine. To fix faster this problem i need to repeat you environment..

Florz replied:
@platinumthinker that depends on which name you are providing. On my WTP inside the wtp.conf I specified `name = "wtp-ap";`.  The hostname is `LEDE` and inside the WTP certificate I am using `BIC-IRAP-WTP-03`.

/v1/wtp returns:
`[
  {
    "id": "���w�7",
    "endpoint": {
      "ip": "none",
      "port": "none"
    }
  },
  {
    "id": "BIC-IRAP-WTP-03",
    "endpoint": {
      "ip": "192.168.1.3",
      "port": 49416
    }
  },
  {
    "id": "���w�8",
    "endpoint": {
      "ip": "none",
      "port": "none"
    }
  }
]`
So this looks like `wtp_id` should be the name from the certificate and you might just be off by one in the data structure. The IDs of the disconnected APs seem to be sort of random but the first one looks familiar.
Using the certificate names is also a good idea because they have to be unique.

Florz replied:
@platinumthinker have you been able to reproduce this problem?

platinumthinker replied:
@Florz, i have wtp router and pc with capwap-ac but i can't reproduce this problem. What "id" you excpect?

platinumthinker replied:
@Florz can you get info about values from certificate using this wtp?

Florz replied:
@platinumthinker I think part of the problem is that I have only one of three WTPs connected. Even in the old API (see previous post), the non connected WTPs are listed with faulty IDs. The connected WTP provides a correct ID `BIC-IRAP-WTP-03`.
I think those disconnected WTPs should not be listed here. They probably show up because I configured them in `/etc/ergw-capwap-node/ergw-capwap-node.config`. In the config they are named accordingly as `<<"BIC-IRAP-WTP-01">>` and `<<"BIC-IRAP-WTP-02">>`.

The big problem now is that our STA is listed under the wrong WTP. As shown before the `wtp-id` is non printable and probably matches one of the non connected WTPs, whereas it should show up under WTP number 3.

platinumthinker replied:
@Florz can you provide ergw-capwap-node.config? And if it possible i want getting you certs file from wtp for inspecting too

Florz replied:
@platinumthinker see [here](https://pastebin.com/WwynQkdB)

platinumthinker replied:
@Florz you have 3 wtp? and what you get from http api request "/api/v1/wtp"?
Another thing: what you have in logs?

platinumthinker replied:
I have some config with several wtp (https://pastebin.com/SG5AqVmg) and not a single wtp shows by commands api v1/v2..

Florz replied:
@platinumthinker I have three WTPs configured in `ergw-capwap-node.config` but only one is connected. The other two WTPs are not connected yet,

```
user@IRAP:~$ curl -X "GET" "http://127.0.0.1:8000/api/v1/wtp"
[{"id":"���w�7","endpoint":{"ip":"none","port":"none"}},{"id":"BIC-IRAP-WTP-03","endpoint":{"ip":"192.168.1.3","port":51765}},{"id":"���w�8","endpoint":{"ip":"none","port":"none"}}]
user@IRAP:~$ curl -X "GET" "http://127.0.0.1:8000/api/v1/station"
[{"id":"���w�7","address":{"ip":"none","port":"none"},"macs":["08-C5-E1-71-94-39"]}]
user@IRAP:~$ curl -X "GET" "http://127.0.0.1:8000/api/v2/station"
[[{"wtp-id":"���w�7","radios":[{"radio-id":1,"stations":[{"mac":"08:c5:e1:71:94:39","rssi":-62,"vlan":0,"rx-packets":122,"rx-bytes":27268,"tx-packets":15,"tx-bytes":9098}]}]}]]
```
The phone is connected to `BIC-IRAP-WTP-03` for sure, as this is our only WTP at the moment. I uploaded my logs [here](https://pastebin.com/yuByvjJh).
I am also in contact with Andrey Teplyashin now, about this possible bug.

Florz replied:
@platinumthinker it becomes even stranger, when I connect the missing WTPs.

with `BIC-IRAP-WTP-02` added:
```
user@IRAP:~$ curl -X "GET" "http://127.0.0.1:8000/api/v1/wtp"
[{"id":"���w�7","endpoint":{"ip":"none","port":"none"}},{"id":"BIC-IRAP-WTP-02","endpoint":{"ip":"192.168.1.2","port":56725}},{"id":"BIC-IRAP-WTP-03","endpoint":{"ip":"192.168.1.3","port":48080}},{"id":"���w�S","endpoint":{"ip":"none","port":"none"}},{"id":"���w�T","endpoint":{"ip":"none","port":"none"}},{"id":"���w�8","endpoint":{"ip":"none","port":"none"}}]

user@IRAP:~$ curl -X "GET" "http://127.0.0.1:8000/api/v2/station"
[[{"wtp-id":"���w�7","radios":[{"radio-id":1,"stations":[{"mac":"08:c5:e1:71:94:39","rssi":-60,"vlan":0,"rx-packets":41,"rx-bytes":8357,"tx-packets":8,"tx-bytes":3956}]}]}]]
```
with all WTPs attached:
```
user@IRAP:~$ curl -X "GET" "http://127.0.0.1:8000/api/v1/wtp"
[{"id":"���w�\u0004","endpoint":{"ip":"none","port":"none"}},{"id":"���w�7","endpoint":{"ip":"none","port":"none"}},{"id":"BIC-IRAP-WTP-02","endpoint":{"ip":"192.168.1.2","port":59793}},{"id":"BIC-IRAP-WTP-03","endpoint":{"ip":"192.168.1.3","port":51795}},{"id":"���w�S","endpoint":{"ip":"none","port":"none"}},{"id":"���w�T","endpoint":{"ip":"none","port":"none"}},{"id":"���w�8","endpoint":{"ip":"none","port":"none"}},{"id":"BIC-IRAP-WTP-01","endpoint":{"ip":"192.168.1.1","port":39676}},{"id":"���w�\u0003","endpoint":{"ip":"none","port":"none"}}]
```
in this case it also messes up the v2 station list:
```
user@IRAP:~$ curl -X "GET" "http://127.0.0.1:8000/api/v1/station"
[{"id":"BIC-IRAP-WTP-02","address":{"ip":"192.168.1.2","port":59793},"macs":["08-C5-E1-71-94-39"]}]
user@IRAP:~$ curl -X "GET" "http://127.0.0.1:8000/api/v2/station"
[[{"wtp-id":"���w�7","radios":[]}]]
```

platinumthinker replied:
fixed this problem in last commit (https://github.com/travelping/capwap/pull/25/commits/adc4fd758208f6a05937c21c9b7aa4d6bcd98f64)

platinumthinker committed:
Ignore BSS when expected WTPs in WTP register names

Currently BSS storing in WTP register as {BSS_NAME, undefined} and WTPs
storing as {WTP_NAME, {ip, port}}.

RoadRunnr commented on src/ieee80211_station.erl:
```diff
@@ -21,7 +21,7 @@
 %% API
 -export([start_link/3, handle_ieee80211_frame/2, handle_ieee802_3_frame/3,
-	 take_over/3, detach/1, delete/1, start_gtk_rekey/4]).
+	 take_over/3, detach/1, delete/1, start_gtk_rekey/4, get_aaa_session/1]).
```
get_aaa_session/1 seems to be unused

RoadRunnr commented on src/capwap_http_api_handler.erl:
```diff
@@ -342,10 +360,10 @@ fmt_ip(IP) ->
-			 radios_in_use = RadiosInUse,
-			 encryption_sub_element = EncSubElem,
-			 sub_elements = SubElements}) ->
+                      max_radios = MaxRadios,
```
white space cleanups are ok, but they should go into a extra commit that touches only white space

RoadRunnr commented on src/capwap_http_api_handler.erl:
```diff
@@ -140,34 +158,40 @@ handle_request_wtp(<<"POST">>, [_, <<"set-ssid">> | _], Req, State) ->
+    case capwap_tools:mac_to_hex(MACStr) of
         MAC when is_binary(MAC) ->
-            R = capwap:detach_station(MAC),
-            {jsx:encode([{result, R}]), Req, State};
```
doesn't that change the return type? Doing so might be ok, I'm just not 100% sure about it.

RoadRunnr commented:
Also, minimal test case for the API might be nice

platinumthinker committed:
Delete whitespaces
